### PR TITLE
Better JEI compat

### PR DESCRIPTION
**Pull request by WaitingIdly**
changes in this PR:
- added the ability to remove non-items and non-fluids from the ingredients side (requires the type of ingredient to be specified)
- adds the ability to remove all ingredients of a given type
- adds the ability to remove all ingredients :trollface: 
- adds the ability to add and remove catalyst items from categories
- adds the ability to remove all catalysts of a specific category
- adds the ability to add and remove JEI descriptions (a special tab in JEI which displays information related to an item)
- adds lang keys for the wiki and an examples file for the above
- methods formerly directly inside `JEI` are now inside subregistries, so eg `mods.jei.hide` must now be `mods.jei.ingredient.hide`

### CODE REVIEW

**Review by brachy84 — APPROVED** [review]
Seems fine. A short migration guide would be useful.

File: src/main/java/com/cleanroommc/groovyscript/compat/mods/jei/Ingredient.java [hunk 1]
```diff
@@ -0,0 +1,252 @@
...
+        if (IngredientHelper.isItem(ingredient)) {
+            addedIngredients.computeIfAbsent(VanillaTypes.ITEM, k -> new ArrayList<>()).addAll(Arrays.asList(ingredient.getMatchingStacks()));
+        } else if (IngredientHelper.isFluid(ingredient)) {
+            addedIngredients.computeIfAbsent(VanillaTypes.FLUID, k -> new ArrayList<>()).add(IngredientHelper.toFluidStack(ingredient));
+        } else {
+            GroovyLog.msg("Error adding items {}", ingredient)
+                    .add("Could not identify the type")
+                    .error()
+                    .post();
+        }
```
Comment (brachy84) [inline src/main/java/com/cleanroommc/groovyscript/compat/mods/jei/Ingredient.java:95]: You can get the ingredient type with `mezz.jei.api.ingredients.IIngredientRegistry.getIngredientType()`

File: src/main/java/com/cleanroommc/groovyscript/compat/mods/jei/Ingredient.java [hunk 2]
```diff
@@ -0,0 +1,252 @@
...
+                    .error()
+                    .post();
+            return;
+        }
+        if (IngredientHelper.isItem(ingredient)) {
+            hiddenIngredients.computeIfAbsent(VanillaTypes.ITEM, k -> new ArrayList<>()).addAll(Arrays.asList(ingredient.getMatchingStacks()));
+        } else if (IngredientHelper.isFluid(ingredient)) {
+            hiddenIngredients.computeIfAbsent(VanillaTypes.FLUID, k -> new ArrayList<>()).add(IngredientHelper.toFluidStack(ingredient));
+        } else {
+            GroovyLog.msg("Error hiding items {}", ingredient)
```
Comment (brachy84) [inline src/main/java/com/cleanroommc/groovyscript/compat/mods/jei/Ingredient.java:139]: same here

File: src/main/java/com/cleanroommc/groovyscript/compat/mods/jei/Description.java [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+    }
+
+    @MethodDescription(type = MethodDescription.Type.ADDITION)
+    public void add(List<IIngredient> target, List<String> description) {
+        addScripted(Pair.of(target, description));
+    }
+
+    @MethodDescription(type = MethodDescription.Type.ADDITION)
+    public void add(List<IIngredient> target, String... description) {
+        addScripted(Pair.of(target, Lists.newArrayList(description)));
```
Comment (brachy84) [inline src/main/java/com/cleanroommc/groovyscript/compat/mods/jei/Description.java:74]: Why not use `Arrays,asList()`?

File: src/main/java/com/cleanroommc/groovyscript/compat/mods/jei/Ingredient.java (hunk 1 above)
Comment (WaitingIdly) [reply src/main/java/com/cleanroommc/groovyscript/compat/mods/jei/Ingredient.java:95]: we dont have access to the `IIngredientRegistry` yet when we want to run this. as an example, trying this code snippet in place of the if/else block generates an NPE
```java
        hiddenIngredients.computeIfAbsent(JeiPlugin.itemRegistry.getIngredientType(ingredient), k -> new ArrayList<>()).addAll(Arrays.asList(ingredient.getMatchingStacks()));
```

**Pull request merged**